WalletWasabi.Tests/UnitTests/BestEffortEndpointConnectorTests.cs

```diff
@@ -0,0 +1,66 @@
+using System;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+using NBitcoin;
+using NBitcoin.Protocol;
+using NBitcoin.Protocol.Behaviors;
+using Xunit;
+
+namespace WalletWasabi.Tests.UnitTests
+{
+	public class BestEffortEndpointConnectorTests
+	{
+		[Fact]
+		public async Task CanConnectWithDifferentModes()
+        {
+            var connector = new BestEffortEndpointConnector();
+            var nodeConnectionParameters = new NodeConnectionParameters();
+            using var nodes = new NodesGroup(Network.TestNet, nodeConnectionParameters);
+
+            async Task Connect(EndPoint endpoint)
+            {
+                using var socket = new Socket(SocketType.Stream, ProtocolType.Tcp);
+                await connector.ConnectSocket(socket, endpoint, nodeConnectionParameters, CancellationToken.None);
+            }
+
+            Exception ex;
+            ex = await Assert.ThrowsAsync<InvalidOperationException>(
+                async () => await Connect(new IPEndPoint(IPAddress.Loopback, 80)));
+            Assert.Equal("The Endpoint connector is configured to allow only Tor endpoints and the '127.0.0.1:80' enpoint is not one", ex.Message);
+            Assert.Equal(BestEffortEndpointConnector.ConnectionMode.OnionServiceOnly, connector.State.Mode);
+
+            ex = await Assert.ThrowsAsync<InvalidOperationException>(
+                async () => await Connect(new DnsEndPoint("nec4kn4ghql7p7an.onion", 80)));
+            Assert.Equal("SocksSettingsBehavior.SocksEndpoint is not set but the connection is expecting using socks proxy", ex.Message);
+            Assert.Equal(BestEffortEndpointConnector.ConnectionMode.OnionServiceOnly, connector.State.Mode);
+
+            // Timing out - Mode changes from OnionServiceOnly -> AllowGoingThroughTorExitNodes
+            connector.State.LastConnectionTime = DateTimeOffset.UtcNow.AddHours(-1);
+
+            ex = await Assert.ThrowsAnyAsync<SocketException>(
+                async () => await Connect(new IPEndPoint(IPAddress.Loopback, 180)));
+            Assert.Contains("refused", ex.Message);
+            Assert.Equal(BestEffortEndpointConnector.ConnectionMode.AllowGoingThroughTorExitNodes, connector.State.Mode);
+
+            // Simulate there is a connection made (so, it is feasible to have more) 
+            connector.IncreaseConnectedNodesCounter();
+            connector.State.LastConnectionTime = DateTimeOffset.UtcNow.AddHours(-1);
+
+            ex = await Assert.ThrowsAnyAsync<SocketException>(
+                async () => await Connect(new IPEndPoint(IPAddress.Loopback, 180)));
+            Assert.Contains("refused", ex.Message);
+            Assert.Equal(BestEffortEndpointConnector.ConnectionMode.AllowGoingThroughTorExitNodes, connector.State.Mode);
+
+            // Simulate there are no connections made (so, go to desperate mode and connect to clearnet) 
```

Comment:
Clearnet or via Tor using exit nodes?

Reply:
Clearnet

Reply:
I don't understand, does it mean that, if a connession to a .onion service or a Tor-to-clearnet is not possible, the Wallet fallback tries a clearnet connession??

Reply:
yes.

Reply:
I don't think this is a good idea.